File: win/TournamentDaemon/TournamentDaemon.cpp
// This is the main DLL file.

#include "TournamentDaemon.h"
#include "../../tournamentd/logger.hpp"

using namespace System::Threading;

TBWin::TournamentDaemon::TournamentDaemon() : port(0), running(false), tourney(new tournament), thread(gcnew Thread(gcnew ThreadStart(this, &TBWin::TournamentDaemon::main)))
{
}

TBWin::TournamentDaemon::~TournamentDaemon()
{
	this->Stop();
	delete tourney;
}

void TBWin::TournamentDaemon::main()
{
	while(running)
	{
		auto quit = tourney->run();
		running = running && !quit;
	}
}

int TBWin::TournamentDaemon::Start(int code)
{
	tourney->authorize(code);
	auto service(tourney->listen(nullptr));

	// server is listening. mark as running and run in background
	running = true;
	thread->IsBackground = true;
	thread->Start();

	// store the port
	port = service.second;

	// return the listening port, for subsequent connection
	return port;
}

void TBWin::TournamentDaemon::Publish(System::String^ name)
{
	// TODO: Stop net service

	if (port)
	{
		// TODO: Start net service
	}
}

void TBWin::TournamentDaemon::Stop()
{
	// TODO: Stop net service

	// clear port
	port = 0;

	// signal thread execution to stop
	running = false;

	// join thread
	thread->Join();
}

File: TBWin/TournamentDaemon.cs
using System.Threading;
using ZeroconfService;

namespace TBWin
{
    class TournamentDaemon
    {
        public int Start(int code)
        {
            // tournament object
            TournamentObject tourney = new TournamentObject();

            // start the service and get the port
            _port = tourney.Start(code);

            // server is listening. mark as running
            _running = true;

            // create a background thread
            _thread = new Thread((ThreadStart)delegate
            {
                while(_running)
                {
                    bool quit = tourney.Run();
                    _running = _running && !quit;
                }
            });
            _thread.IsBackground = true;
            _thread.Start();

            // return the listening port, for subsequent connection
            return _port;
        }

        public bool Publish(string name)
        {
            // stop net service
            if(_netService != null)
            {
                _netService.Stop();
                _netService = null;
            }

            // start new net service and publish
            if(_port != 0)
            {
                _netService = new NetService("local.", "_tournbuddy._tcp.", name, _port);
                try
                {
                    _netService.Publish();
                    return true;
                }
                catch(System.DllNotFoundException)
                {
                    // bonjour not installed. just return false to indicate we did not publish
                }
                _netService = null;
            }

            return false;
        }

        public void Stop()
        {
            // stop net service
            if (_netService != null)
            {
                _netService.Stop();
            }

            // clear port
            _port = 0;

            // signal thread execution to stop
            _running = false;

            // join thread
            _thread.Join();
        }

        // Port in use
        private int _port;

        // Flag to control whether background thread is running
        private bool _running;

        // Thread
        private Thread _thread;

        // Net service
        private NetService _netService;
    }
}
